### PR TITLE
Testing Java Projects with JUnit and Mockito- Doing more with Mockito

### LINKED ISSUE #6138: [Languages] Testing Java Projects using JUnit and Mockito - Doing more with Mockito
#### Proposed title of article

[Languages] Testing Java Projects using JUnit and Mockito - Doing more with Mockito

#### Proposed article introduction

In the real world, software has dependencies. We usually have services that depend on a database or some other third party service. The concept of mocking enables us to verify that our code works without addressing concerns about dependencies. The theory is that if my code works as designed and the external dependencies work as designed, then when integrated my system should work as designed. Mockito is a mocking framework in Java. It can be integrated with Java's testing framework JUnit. Mockito allows us to mock objects whose behaviour we can specify within our test. This article is intended to be a sequel to an earlier article also published by the engineering program under the topic- [Testing Java Projects using JUnit and Mockito](https://www.section.io/engineering-education/mocking-with-junit-and-mockito-the-why-and-how/). In this new article, I intend to demonstrate to my readers:
- how to use Mockito Argument Matchers
- how to write mock tests for void methods
- how to throw exceptions with Mockito
- how to work with inline stubbing
- how to query mock objects for their details

#### Key takeaways 

At the end of this tutorial, the reader should:
- understand the use of Mockito's Argument Matchers and know when to apply them
- be able to write mock tests for void methods
- be able to throw exceptions with Mockito
- be able to query mock objects for their details
- be able to work with inline stubbing

#### Article quality

This article is intended for persons new to mock testing and can also be useful to experienced programmers. The theory behind mocking and stubbing is so vast and can be confusing for newbies to unit testing. In addition to my earlier article, this article offers more content on mock testing using Mockito than eighty per cent of other blogs and posts about Mockito.
Finally, my article demonstrates how to perform mocking in unit tests while building a Hostel Management System using Java, Mockito and Java spark framework. A Hostel Management System is a common search topic among Java students with over six million searches on Google. This article is guaranteed to increase my reader's proficiency in using Mockito.

#### References

- https://site.mockito.org/
- [Java Argument Matchers Doc](https://javadoc.io/static/org.mockito/mockito-core/3.3.3/org/mockito/ArgumentMatchers.html)
- Mastering Unit Testing Using Mockito and JUnit- An advanced guide to mastering unit testing using Mockito and JUnit by Sujoy Acharya.
- [Mockito ArgumentMatchers](https://www.baeldung.com/mockito-argument-matchers).

### PR DESCRIPTION
**Pull request by ehizman**
## Our **Peer Reviewer and Content Moderator teams do NOT provide any revisions services.** All revisions and edits should be completed on your own forked repo (as to not take up room in the queue) in order for our team to review them all in a timely manner. 

**ANY ARTICLE SUBMITTED WITH GLARING ERRORS WILL BE IMMEDIATELY CLOSED.**

#### Formatting and structure checks
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**

- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Is the article you are submitting an in-depth and unique article? Does it go beyond what is in the official docs and what is covered in other blog sites. See these articles as [examples](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md).
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**

- [x] Have you included a hero image that is **600x400 pixels**, under **300KB in size**, and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
      
#### Grammar & plagiarism checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) - [Grammarly](https://grammarly.com) 
- [x] Please place your article through a 3rd party plagiarism checker. We suggest using Quetext, [this tool](https://www.quetext.com/) is free and has a daily limit. **We typically accept articles with 10% or less.**
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [x] Have you added sources for quotes and images that aren't yours?

#### Technical checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**
- [x] Have you checked your code runs correctly and you've highlighted all necessary dependencies for installation?
- [x] Are the software programs and packages you're highlighting in your article up to date, using current versions and not deprecated?

This closes #6138

### CODE REVIEW

**Comment by srishilesh** [comment]
Hello @ehizman 

Thanks for submitting this PR.

Your PR has entered our reviewing queue. Please wait until a peer-reviewer pings you to get started with the review.

Meanwhile, we suggest you to go through approved PRs/published articles to see if you have missed out anything. This would help us avoid any potential closure of the PR.

Thanks.

**Comment by ehizman** [comment]
> Hello @ehizman
> 
> Thanks for submitting this PR.
> 
> Your PR has entered our reviewing queue. Please wait until a peer-reviewer pings you to get started with the review.
> 
> Meanwhile, we suggest you to go through approved PRs/published articles to see if you have missed out anything. This would help us avoid any potential closure of the PR.
> 
> Thanks.

**Comment by ehizman** [comment]
Hello @srishilesh 
It's been days now and my PR has not been attended to.
???

**Comment by srishilesh** [comment]
Hey @ehizman ,

The peer-reviewers are working towards completing existing reviews to take up more. Please wait patiently until a peer-reviewer pings you to get started.

Thanks.

**Comment by srishilesh** [comment]
Hey @ehizman ,

I'll get started with the preliminary review by EOD.

Thanks.

**Review by srishilesh — CHANGES_REQUESTED** [review]
Hey @ehizman ,

On going through your article, I found several instances where it requires explanation of code snippets. I've pointed out a few.

Moreover, this tutorial is a little low on content. I believe on addressing my suggestions, it should be good to go.

Also, I found several instances where the content was repetitive. Always make sure to re-read your article once you finish it.

As you know that, any PRs that requires major revisions will be closed. So, you make work on these changes locally, and create a new PR when done. Ping me in the new PR.

Let me know if you have any questions.

Thanks.

File: content/articles/testing-java-projects-using-junit-and-mockito-doing-more-with-mockito/index.md [hunk 1]
```diff
@@ -0,0 +1,180 @@
+### Introduction
+
+How do we know that code works except by testing. Various forms of testing exist in software engineering. [Unit testing](https://en.wikipedia.org/wiki/Unit_testing) is the most vital form of testing. Unit testing offers a means to test the individual code components as isolated units. The key in unit testing is the isolation of program units. Programmers use stubs, spies and mock objects to mock out external dependencies. The use of stubs, spies and mock objects enable code isolation.
+
+
+Mockito is the most popular mocking framework for Java projects today. Mockito offers intuitive matchers that allow programmers to mock the behaviours of dependencies.
+
+The aim of this tutorial is to guide readers on how to write effective units tests using Mockito and JUnit. This tutorial is a continuation of [Testing Java Projects using JUnit and Mockito](https://www.section.io/engineering-education/mocking-with-junit-and-mockito-the-why-and-how/), in which I provided an introduction to Mockito and guided readers on how to integrate Mockito into a Java project amongst other things. I recommend that you read the earlier article before continuing.
```
[the remark this answers is not in the document]
Comment (ehizman) [reply content/articles/testing-java-projects-using-junit-and-mockito-doing-more-with-mockito/index.md:8]: What if I like a personal touch to my writing?

[the remark this answers is not in the document]
Comment (srishilesh) [reply content/articles/testing-java-projects-using-junit-and-mockito-doing-more-with-mockito/index.md:8]: @ehizman 
I totally get it. Here, at Section, it's a bit more formal. And, to maintain standards, we suggest this.

Alternatively, you can use "we", "us", "let's" and so on. You can take a look at other articles to see how they've done it.

**Pull request closed without merging**